Mathlib/GroupTheory/DivisibleHull.lean

```diff
@@ -0,0 +1,416 @@
+
+end Group
+
+section LinearOrder
+variable {M : Type*} [AddCommMonoid M] [LinearOrder M] [IsOrderedCancelAddMonoid M]
+
+private theorem lift_aux (m n m' n' : M) (s t s' t' : ℕ+)
+    (h : mk m s = mk m' s') (h' : mk n t = mk n' t') :
+    (t.val • m ≤ s.val • n) = (t'.val • m' ≤ s'.val • n') := by
+  rw [mk_eq_mk_iff_smul_eq_smul] at h h'
+  rw [propext_iff, ← (nsmul_right_strictMono (mul_ne_zero s'.ne_zero t'.ne_zero)).le_iff_le]
+  convert (nsmul_right_strictMono (M := M) (mul_ne_zero s.ne_zero t.ne_zero)).le_iff_le using 2
+  · simp_rw [smul_smul, mul_rotate s'.val, ← smul_smul, h, smul_smul]
+    ring_nf
+  · simp_rw [smul_smul, ← mul_rotate s'.val, ← smul_smul, ← h', smul_smul]
+    ring_nf
+instance : LE (DivisibleHull M) where
+  le x y := liftOn₂ x y (fun m s n t ↦ t.val • m ≤ s.val • n) lift_aux
+
+theorem mk_le_mk {m m' : M} {s s' : ℕ+} :
+    mk m s ≤ mk m' s' ↔ s'.val • m ≤ s.val • m' := by rfl
+
+instance : LinearOrder (DivisibleHull M) where
+  le_refl a := by
+    induction a with | mk m s
+    rw [mk_le_mk]
+  le_trans a b c hab hbc := by
+    induction a with | mk ma sa
+    induction b with | mk mb sb
+    induction c with | mk mc sc
+    rw [mk_le_mk] at ⊢ hab hbc
+    rw [← (nsmul_right_strictMono (show sb.val ≠ 0 by simp)).le_iff_le, smul_comm _ _ ma,
+      smul_comm _ _ mc]
+    rw [← (nsmul_right_strictMono (show sc.val ≠ 0 by simp)).le_iff_le, smul_comm _ _ mb] at hab
+    rw [← (nsmul_right_strictMono (show sa.val ≠ 0 by simp)).le_iff_le] at hbc
+    exact hab.trans hbc
+  le_antisymm a b h h' := by
+    induction a with | mk ma sa
+    induction b with | mk mb sb
+    rw [mk_le_mk] at h h'
+    rw [mk_eq_mk_iff_smul_eq_smul]
+    exact le_antisymm h h'
+  le_total a b := by
+    induction a with | mk ma sa
+    induction b with | mk mb sb
+    simp_rw [mk_le_mk]
+    exact le_total _ _
+  toDecidableLE := by
+    unfold DecidableLE LE.le instLE liftOn₂ LocalizedModule.liftOn₂
+    infer_instance
+
+theorem mk_lt_mk {m m' : M} {s s' : ℕ+} : mk m s < mk m' s' ↔ s'.val • m < s.val • m' := by
+  simp_rw [lt_iff_not_ge, mk_le_mk]
+
+-- An intermediate lemma for `IsOrderedCancelAddMonoid`.
+-- For public API, use `_root_.add_lt_add_left`.
+private lemma aux_add_lt_add_left (a : DivisibleHull M) {b c : DivisibleHull M} (h : b < c) :
+    a + b < a + c := by
+  induction a with | mk ma sa
+  induction b with | mk mb sb
+  induction c with | mk mc sc
+  simp_rw [mk_add_mk]
+  rw [mk_lt_mk] at ⊢ h
+  simp_rw [PNat.mul_coe, mul_smul, smul_add, smul_smul]
+  have := add_lt_add_left (nsmul_lt_nsmul_right (sa * sa).ne_zero h) ((sa * sb * sc.val) • ma)
+  simp_rw [PNat.mul_coe, smul_smul] at this
+  convert this using 3 <;> ring
+
+instance : IsOrderedCancelAddMonoid (DivisibleHull M) where
+  add_le_add_left a b h c := by
+    obtain rfl | h := eq_or_lt_of_le h
+    · simp
+    exact (aux_add_lt_add_left c h).le
+  le_of_add_le_add_left a b c h := by
+    contrapose! h
+    exact aux_add_lt_add_left a h
```

Comment:
Sounds like we are missing a constructor for `IsOrderedCancelAddMonoid`! Mind adding and using it?